### PR TITLE
Add FFT Ops

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves https://github.com/keras-team/keras-core/issues/429

The reason for not supporting all args at the moment is that TensorFlow supports none of those args. Moreover, there are some inconsistencies between the NumPy FFT API and JAX FFT API. We need a very minimal version for FNet (and for most usecases, I suppose). This can be extended later to support all args.

### CODE REVIEW

File: keras_core/backend/tensorflow/math.py [hunk 1]
```diff
@@ -33,3 +35,74 @@ def qr(x, mode="reduced"):
     if mode == "reduced":
         return tf.linalg.qr(x)
     return tf.linalg.qr(x, full_matrices=True)
+
+
+def _get_complex_tensor_from_tuple(a):
```
Comment (abheesht17) [inline keras_core/backend/tensorflow/math.py:40]: Let me know if "helper" functions are to be moved to some other file.

File: keras_core/backend/jax/math.py [hunk 2]
```diff
@@ -43,3 +43,73 @@ def qr(x, mode="reduced"):
...
+            "`norm` argument value not supported. "
+            f"Expected `None`. Received: norm={norm}"
+        )
+    complex_input = _get_complex_tensor_from_tuple(a)
+    complex_output = jax.numpy.fft.fft(complex_input)
+    return jax.numpy.real(complex_output), jax.numpy.imag(complex_output)
+
+
+def fft2(a, s=None, axes=(-2, -1), norm=None):
+    if s is not None:
```
Comment (mattdangerw) [inline keras_core/backend/jax/math.py:98]: might be worth dropping a comment here, tensorflow does not support `s` and `axes` options, so we will hold off on support for these options for now.

File: keras_core/backend/jax/math.py [hunk 3]
```diff
@@ -43,3 +43,73 @@ def qr(x, mode="reduced"):
             f"Received: mode={mode}"
         )
     return jax.numpy.linalg.qr(x, mode=mode)
+
+
+def _get_complex_tensor_from_tuple(a):
+    if not isinstance(a, tuple):
```
Comment (mattdangerw) [inline keras_core/backend/jax/math.py:49]: maybe `not isinstance(a, tuple) or len(a) != 2`?

Comment (mattdangerw) [reply keras_core/backend/jax/math.py:49]: also probably support list or tuple, generally interchangeable as input in keras/keras-core

**Review by mattdangerw — APPROVED** [review]
This looks good to me, just a few comments

File: keras_core/backend/jax/math.py [hunk 4]
```diff
@@ -43,3 +43,77 @@ def qr(x, mode="reduced"):
...
+            f"Received: a={a}"
+        )
+    # `convert_to_tensor` does not support passing complex tensors. We separate
+    # the input out into real and imaginary and convert them separately.
+    real, imag = a
+    # Check shapes.
+    if real.shape != imag.shape:
+        raise ValueError(
+            "Input `a` should be a tuple of two tensors - real and imaginary."
+            "Both real and imaginary should have the same shape. "
```
Comment (fchollet) [inline keras_core/backend/jax/math.py:61]: "Both the real and imaginary parts..."

File: keras_core/backend/jax/math.py [hunk 5]
```diff
@@ -43,3 +43,77 @@ def qr(x, mode="reduced"):
...
+            f"Received: real.shape = {real.shape}, imag.shape = {imag.shape}"
```
Comment (fchollet) [inline keras_core/backend/jax/math.py:62]: refer to them as a[0], a[1]

File: keras_core/ops/math.py [hunk 6]
```diff
@@ -155,3 +155,143 @@ def qr(x, mode="reduced"):
...
+            KerasTensor(shape=real.shape, dtype=real.dtype),
+            KerasTensor(shape=imag.shape, dtype=imag.dtype),
+        )
+
+    def call(self, x):
+        return backend.math.fft2(x, s=self.s, axes=self.axes, norm=self.norm)
+
+
+@keras_core_export("keras_core.ops.fft")
+def fft(a, n=None, axis=-1, norm=None):
```
Comment (fchollet) [inline keras_core/ops/math.py:287]: Add docstrings to these 2 symbols.

File: keras_core/backend/torch/math.py [hunk 7]
```diff
@@ -81,3 +81,78 @@ def qr(x, mode="reduced"):
...
+        )
+
+    complex_input = torch.complex(real, imag)
+    return complex_input
+
+
+def fft(a, n=None, axis=-1, norm=None):
+    # TF does not support the following args. Hence, we are not supporting
+    # them right now.
+    if n is not None:
```
Comment (fchollet) [inline keras_core/backend/torch/math.py:118]: My general take is that if we do not support these arguments then they should not be in the signature. The values used should be documented in the public-facing docstrings (for the `keras_core.ops` functions).

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by fchollet** [comment]
Are you able to add it to the numpy backend as well? We can probably just redirect to jax like we do for most nn ops.

**Comment by abheesht17** [comment]
> Are you able to add it to the numpy backend as well? We can probably just redirect to jax like we do for most nn ops.

Redirected to JAX for NumPy backend!

**Review by fchollet — APPROVED** [review]
LGTM -- thanks for the great contribution! 👍

**Pull request merged**